### PR TITLE
Make `num_classes` optional, in case of micro averaging

### LINKED ISSUE #2837: `num_classes` should not be required when using micro averaging
## 🐛 Bug

When using micro averaging on multiclass_accuracy, it is required to pass in `num_classes`, even though it's not part of the calculation.

### To Reproduce

```py
from torchmetrics.functional.classification import multiclass_accuracy
import torch
x = torch.Tensor([0, 0, 0])
y = torch.Tensor([0, 1, 1])

print(multiclass_accuracy(x, y, average="macro", num_classes=2))
print(multiclass_accuracy(x, y, num_classes=2))
print(multiclass_accuracy(x, y, average="micro", num_classes=2))
try:
  print(multiclass_accuracy(x, y, average="micro"))
except:
  print("Need num_classes even for micro, why?")
```

### Expected behavior

In general, I expect the metric to be computed.  In the above code sample, I expect 0.33 to be returned, rather than an exception.

### Environment

- TorchMetrics version: 1.6.0
- Python version: 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0]
- PyTorch version: 2.5.0+cu121
- Colab https://colab.research.google.com/drive/1tVJJkktvEmRJgKGgzKrAsKYSx2c_hjSs?usp=sharing

### PR DESCRIPTION
**Pull request by baskrahmer**
## What does this PR do?

Fixes #2837

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--2841.org.readthedocs.build/en/2841/

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@baskrahmer how much is missing here?

**Comment by baskrahmer** [comment]
> @baskrahmer how much is missing here?

I just need to check why the CI is failing.

**Review by lantiga — APPROVED** [review]
Thanks!

**Comment by lantiga** [comment]
Actually, shouldn't we check whether `num_classes` is not passed when microaveraging is not `None` and produce an informative error in that case?

File: src/torchmetrics/classification/stat_scores.py [hunk 1]
```diff
@@ -307,7 +307,7 @@ class MulticlassStatScores(_AbstractStatScores):
 
     def __init__(
         self,
-        num_classes: int,
+        num_classes: Optional[int] = None,
```
Comment (Borda) [inline src/torchmetrics/classification/stat_scores.py:310]: @baskrahmer let's check https://github.com/Lightning-AI/torchmetrics/pull/2841#issuecomment-2558054273
